### PR TITLE
[CORE-571] Gov: Add New Market e2e test

### PR DESCRIPTION
**Pull request by teddyding**

### CODE REVIEW

File: protocol/testutil/delaymsg/delaymsg_helpers.go [hunk 1]
```diff
@@ -15,10 +15,3 @@ func CreateTestAnyMsg(t *testing.T) *codectypes.Any {
 	require.NoError(t, err)
 	return any
 }
-
-// EncodeMessageToAny converts a message to an Any object for protobuf encoding.
-func EncodeMessageToAny(t *testing.T, msg sdk.Msg) *codectypes.Any {
```
Comment (teddyding) [inline protocol/testutil/delaymsg/delaymsg_helpers.go:20]: Moved to `testutil/encoding`

File: protocol/testutil/app/gov.go [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+var (
+	TestVotingPeriod = 1 * time.Minute
+)
+
+// SubmitAndTallyProposal simulates the following:
+//   - A proposal with the given messages is submitted. Check proposal submission succeeds or fails as expected.
+//   - If proposal successfully submitted:
+//     -- All validators vote for the proposal.
+//     --The proposal is tallied after voting period ends.
+func SubmitAndTallyProposal(
```
Comment (teddyding) [inline protocol/testutil/app/gov.go:32]: @tqin7 This function can probablys be reused for testing `Gov - Param Updates`

File: protocol/testutil/tx/msg_tx.go [hunk 3]
```diff
@@ -26,16 +24,11 @@ func MustGetTxBytes(msgs ...sdk.Msg) []byte {
...
-		return v.Order.OrderId.SubaccountId.Owner
-	case *clobtypes.MsgCancelOrder:
-		return v.OrderId.SubaccountId.Owner
-	case *sendingtypes.MsgDepositToSubaccount:
-		return v.Sender
-	case *sendingtypes.MsgWithdrawFromSubaccount:
-		return v.Sender.Owner
-	default:
-		panic(fmt.Errorf("Not a supported type %T", msg))
+	if len(msg.GetSigners()) == 0 {
```
Comment (teddyding) [inline protocol/testutil/tx/msg_tx.go:27]: Updated with a generic implementation

File: protocol/testutil/app/gov.go [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+
+	proposals := tApp.App.GovKeeper.GetProposals(ctx)
+	require.Len(t, proposals, 1)
+
+	// Have all 4 validators vote for the proposal.
+	for _, validator := range []sdk.AccAddress{
+		constants.AliceAccAddress,
+		constants.BobAccAddress,
+		constants.CarlAccAddress,
+		constants.DaveAccAddress,
```
Comment (tqin7) [inline protocol/testutil/app/gov.go:94]: I wonder if we can initialize the app with some kind of staking distribution so that this doesn't have to be hard-coded. I'll look into it as part of my PRs

File: protocol/lib/bank.go [hunk 5]
```diff
@@ -0,0 +1,13 @@
+package lib
+
+import (
+	sdk "github.com/cosmos/cosmos-sdk/types"
+)
+
+func MustParseCoinsNormalized(coinStr string) sdk.Coins {
```
Comment (tqin7) [inline protocol/lib/bank.go:7]: nit: add some test cases for this

File: protocol/testing/e2e/gov/add_new_market_test.go [hunk 6]
```diff
@@ -0,0 +1,256 @@
...
+				return
+			}
+
+			switch tc.expectedProposalStatus {
+			case govtypesv1.ProposalStatus_PROPOSAL_STATUS_FAILED:
+				require.Equal(t, initMarketParams, tApp.App.PricesKeeper.GetAllMarketParams(ctx))
+				require.Equal(t, initPerpetuals, tApp.App.PerpetualsKeeper.GetAllPerpetuals(ctx))
+				require.Equal(t, initClobPairs, tApp.App.ClobKeeper.GetAllClobPairs(ctx))
+			case govtypesv1.ProposalStatus_PROPOSAL_STATUS_PASSED:
+				// Proposal passed and and succesfully executed, check states are updated.
```
Comment (tqin7) [inline protocol/testing/e2e/gov/add_new_market_test.go:222]:
```suggestion
				// Proposal passed and successfully executed, check states are updated.
```

File: protocol/testutil/app/gov.go [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+	require.NoError(t, err)
+
+	// Create a signed transaction with MsgSubmitProposal
+	submitProposalCheckTx := MustMakeCheckTxWithPrivKeySupplier(
+		ctx,
+		tApp.App,
+		MustMakeCheckTxOptions{
+			AccAddressForSigning: constants.AliceAccAddress.String(),
+			Gas:                  constants.TestGasLimit,
+			FeeAmt:               constants.TestFeeCoins_5Cents,
```
Comment (tqin7) [inline protocol/testutil/app/gov.go:58]: for my understanding: does CheckTx fail if we don't pass in a fee?

**Review by tqin7 — APPROVED** [review]

File: protocol/lib/bank.go (hunk 5 above)
Comment (ttl33) [reply protocol/lib/bank.go:7]: nit: should we add a quick doc comment (i.e. let CoPilot draft one?)

File: protocol/testing/e2e/gov/add_new_market_test.go [hunk 8]
```diff
@@ -0,0 +1,256 @@
...
+	clobtypes "github.com/dydxprotocol/v4-chain/protocol/x/clob/types"
+	delaymsgtypes "github.com/dydxprotocol/v4-chain/protocol/x/delaymsg/types"
+	perptypes "github.com/dydxprotocol/v4-chain/protocol/x/perpetuals/types"
+	pricestypes "github.com/dydxprotocol/v4-chain/protocol/x/prices/types"
+	"github.com/stretchr/testify/require"
+)
+
+func TestAddNewMarketProposal(t *testing.T) {
+	testMarketParam1001 := pricestest.GenerateMarketParamPrice(
+		pricestest.WithId(1001),
```
Comment (ttl33) [inline protocol/testing/e2e/gov/add_new_market_test.go:25]: nit: define a constant for ID and use it across prices, perps, clob modules

File: protocol/testing/e2e/gov/add_new_market_test.go [hunk 9]
```diff
@@ -0,0 +1,256 @@
...
+				&delaymsgtypes.MsgDelayMessage{
+					Authority:   authtypes.NewModuleAddress(govtypes.ModuleName).String(),
+					Msg:         encoding.EncodeMessageToAny(t, msgUpdateClobPair1001ToActive),
+					DelayBlocks: 10,
+				},
+			},
+			updateClobDelayBlocks:  10,
+			expectedProposalStatus: govtypesv1.ProposalStatus_PROPOSAL_STATUS_PASSED,
+		},
+		"Proposals execution fails due to incorrectly ordered messages": {
```
Comment (ttl33) [inline protocol/testing/e2e/gov/add_new_market_test.go:74]: nit: can we use the following test name convention?

```
"Success: with 4 valid messages"
"Fail: ..."
```

File: protocol/testing/e2e/gov/add_new_market_test.go [hunk 10]
```diff
@@ -0,0 +1,256 @@
...
+				t,
+				ctx,
+				&tApp,
+				tc.proposedMsgs,
+				tc.expectSubmitProposalFail,
+				tc.expectedProposalStatus,
+			)
+
+			if tc.expectSubmitProposalFail {
+				return
```
Comment (ttl33) [inline protocol/testing/e2e/gov/add_new_market_test.go:213]: if the proposal submission fails, are there conditions that we assert here?

like there's no pending gov proposal, etc

File: protocol/testing/e2e/gov/add_new_market_test.go [hunk 11]
```diff
@@ -0,0 +1,256 @@
...
+				// Check market
+				marketParam, exists := tApp.App.PricesKeeper.GetMarketParam(ctx, testMarketParam1001.Param.Id)
+				require.True(t, exists)
+				require.Equal(t, testMarketParam1001.Param, marketParam)
+
+				marketPrice, err := tApp.App.PricesKeeper.GetMarketPrice(ctx, testMarketParam1001.Param.Id)
+				require.NoError(t, err)
+				require.Equal(t, pricestypes.MarketPrice{
+					Id:       testMarketParam1001.Param.Id,
+					Price:    0,
```
Comment (ttl33) [inline protocol/testing/e2e/gov/add_new_market_test.go:232]: nit: add comment why this is 0

File: protocol/testing/e2e/gov/add_new_market_test.go [hunk 12]
```diff
@@ -0,0 +1,256 @@
...
+					Exponent: testMarketParam1001.Param.Exponent,
+				}, marketPrice)
+				// Check perpeutal
```
Comment (ttl33) [inline protocol/testing/e2e/gov/add_new_market_test.go:235]: nit: add new line above

File: protocol/testing/e2e/gov/add_new_market_test.go [hunk 13]
```diff
@@ -0,0 +1,256 @@
...
+
+				// Check clob
+				clobPair, exists := tApp.App.ClobKeeper.GetClobPair(ctx, clobtypes.ClobPairId(testClobPair1001.Id))
+				require.True(t, exists)
+				require.Equal(t, clobPair, *testClobPair1001)
+
+				ctx = tApp.AdvanceToBlock(uint32(ctx.BlockHeight())+tc.updateClobDelayBlocks+1, testapp.AdvanceToBlockOptions{})
+				clobPair, exists = tApp.App.ClobKeeper.GetClobPair(ctx, clobtypes.ClobPairId(testClobPair1001.Id))
+				require.True(t, exists)
+				// Check that clob pair is updated.
```
Comment (ttl33) [inline protocol/testing/e2e/gov/add_new_market_test.go:248]: nit: add new line above

File: protocol/testing/e2e/gov/add_new_market_test.go [hunk 14]
```diff
@@ -0,0 +1,256 @@
...
+				}, marketPrice)
+				// Check perpeutal
+				perp, err := tApp.App.PerpetualsKeeper.GetPerpetual(ctx, testPerpetual1001.Params.Id)
+				require.NoError(t, err)
+				require.Equal(t, testPerpetual1001.Params, perp.Params)
+
+				// Check clob
+				clobPair, exists := tApp.App.ClobKeeper.GetClobPair(ctx, clobtypes.ClobPairId(testClobPair1001.Id))
+				require.True(t, exists)
+				require.Equal(t, clobPair, *testClobPair1001)
```
Comment (ttl33) [inline protocol/testing/e2e/gov/add_new_market_test.go:243]: nit: 
```suggestion
				require.Equal(t, *testClobPair1001, clobPair)
```

File: protocol/testing/e2e/gov/add_new_market_test.go [hunk 15]
```diff
@@ -0,0 +1,256 @@
...
+					DelayBlocks: 10,
+				},
+			},
+			updateClobDelayBlocks:  10,
+			expectedProposalStatus: govtypesv1.ProposalStatus_PROPOSAL_STATUS_FAILED,
+		},
+		"Proposal submission fails, due to invalid signer": {
+			proposedMsgs: []sdk.Msg{
+				&pricestypes.MsgCreateOracleMarket{
+					Authority: authtypes.NewModuleAddress(clobtypes.ModuleName).String(),
```
Comment (ttl33) [inline protocol/testing/e2e/gov/add_new_market_test.go:130]: nit: add comment that the authority here is invalid

File: protocol/testing/e2e/gov/add_new_market_test.go [hunk 16]
```diff
@@ -0,0 +1,256 @@
...
+			},
+			updateClobDelayBlocks:  10,
+			expectedProposalStatus: govtypesv1.ProposalStatus_PROPOSAL_STATUS_FAILED,
+		},
+		"Proposal execution fails due to existing objects": {
+			proposedMsgs: []sdk.Msg{
+				&pricestypes.MsgCreateOracleMarket{
+					Authority: authtypes.NewModuleAddress(govtypes.ModuleName).String(),
+					Params: pricestest.GenerateMarketParamPrice(
+						pricestest.WithId(5),
```
Comment (ttl33) [inline protocol/testing/e2e/gov/add_new_market_test.go:103]: nit: add comment that this ID already exists

File: protocol/testing/e2e/gov/add_new_market_test.go [hunk 17]
```diff
@@ -0,0 +1,256 @@
...
+		proposedMsgs             []sdk.Msg
+		updateClobDelayBlocks    uint32
+		expectSubmitProposalFail bool
+		expectedProposalStatus   govtypesv1.ProposalStatus
+	}{
+		"Success with 4 standard messages": {
+			proposedMsgs: []sdk.Msg{
+				&pricestypes.MsgCreateOracleMarket{
+					Authority: authtypes.NewModuleAddress(govtypes.ModuleName).String(),
+					Params:    testMarketParam1001.Param,
```
Comment (ttl33) [inline protocol/testing/e2e/gov/add_new_market_test.go:55]: for my own understanding, is 1001 ID okay because we no longer require the IDs to be consecutive?

File: protocol/testing/e2e/gov/add_new_market_test.go [hunk 18]
```diff
@@ -0,0 +1,256 @@
...
+					Authority: authtypes.NewModuleAddress(govtypes.ModuleName).String(),
+					Params:    testPerpetual1001.Params,
+				},
+				&delaymsgtypes.MsgDelayMessage{
+					Authority:   authtypes.NewModuleAddress(govtypes.ModuleName).String(),
+					Msg:         encoding.EncodeMessageToAny(t, msgUpdateClobPair1001ToActive),
+					DelayBlocks: 10,
+				},
+			},
+			updateClobDelayBlocks:  10,
```
Comment (ttl33) [inline protocol/testing/e2e/gov/add_new_market_test.go:95]: nit: can we just use the delayblock value in the above `MsgDelayMessage`? why do we need to explicitly define this?

File: protocol/testing/e2e/gov/add_new_market_test.go [hunk 19]
```diff
@@ -0,0 +1,256 @@
...
+			expectedProposalStatus: govtypesv1.ProposalStatus_PROPOSAL_STATUS_FAILED,
+		},
+		"Proposal submission fails, due to invalid signer": {
+			proposedMsgs: []sdk.Msg{
+				&pricestypes.MsgCreateOracleMarket{
+					Authority: authtypes.NewModuleAddress(clobtypes.ModuleName).String(),
+					Params: pricestest.GenerateMarketParamPrice(
+						pricestest.WithId(5),
```
Comment (ttl33) [inline protocol/testing/e2e/gov/add_new_market_test.go:132]: nit: should these be valid IDs for this failure test case and below?

File: protocol/testing/e2e/gov/add_new_market_test.go [hunk 20]
```diff
@@ -0,0 +1,256 @@
...
+					Params:    testPerpetual1001.Params,
+				},
+				&clobtypes.MsgCreateClobPair{
+					Authority: authtypes.NewModuleAddress(govtypes.ModuleName).String(),
+					ClobPair:  *testClobPair1001,
+				},
+				&delaymsgtypes.MsgDelayMessage{
+					Authority:   authtypes.NewModuleAddress(govtypes.ModuleName).String(),
+					Msg:         encoding.EncodeMessageToAny(t, msgUpdateClobPair1001ToActive),
+					DelayBlocks: 10,
```
Comment (ttl33) [inline protocol/testing/e2e/gov/add_new_market_test.go:68]: can we add a test case with different DelayBlocks value? specifically

* DelayBlocks: 0
* DelayBlocks: 1

File: protocol/testutil/app/app.go [hunk 21]
```diff
@@ -703,7 +710,7 @@ func MustMakeCheckTxWithPrivKeySupplier(
 		rand.New(rand.NewSource(42)),
 		app.TxConfig(),
 		messages,
-		sdk.Coins{},
+		options.FeeAmt,
```
Comment (ttl33) [inline protocol/testutil/app/app.go:713]: nit: why is this needed?

File: protocol/testutil/app/gov.go [hunk 22]
```diff
@@ -0,0 +1,127 @@
...
+	ctx sdk.Context,
+	tApp *TestApp,
+	messages []sdk.Msg,
+	expectSubmitProposalFails bool,
+	expectedProposalStatus govtypesv1.ProposalStatus,
+) sdk.Context {
+	// Create a MsgSubmitProposal
+	msgSubmitProposal, err := govtypesv1.NewMsgSubmitProposal(
+		messages,
+		sdk.Coins{sdk.NewInt64Coin(constants.TestNativeTokenDenom, 10_000_000)},
```
Comment (ttl33) [inline protocol/testutil/app/gov.go:43]: nit: define a constant for `10_000_000`

File: protocol/testutil/tx/msg_tx.go [hunk 23]
```diff
@@ -26,16 +24,11 @@ func MustGetTxBytes(msgs ...sdk.Msg) []byte {
 
 // Returns the account address that should sign the msg. Will panic if it is an unsupported message type.
 func MustGetSignerAddress(msg sdk.Msg) string {
```
Comment (ttl33) [inline protocol/testutil/tx/msg_tx.go:26]:
```suggestion
func MustGetOnlySignerAddress(msg sdk.Msg) string {
```

**Review by ttl33 — APPROVED** [review]

File: protocol/testing/e2e/gov/add_new_market_test.go [hunk 24]
```diff
@@ -0,0 +1,256 @@
...
+				},
+				&clobtypes.MsgCreateClobPair{
+					Authority: authtypes.NewModuleAddress(govtypes.ModuleName).String(),
+					ClobPair: *clobtest.GenerateClobPair(
+						clobtest.WithId(5),
+					),
+				},
+				&delaymsgtypes.MsgDelayMessage{
+					Authority:   authtypes.NewModuleAddress(govtypes.ModuleName).String(),
+					Msg:         encoding.EncodeMessageToAny(t, msgUpdateClobPair1001ToActive),
```
Comment (ttl33) [inline protocol/testing/e2e/gov/add_new_market_test.go:120]: Should we add a test case where the delay msg is incorrect so although the gov passed, the CLOB market never gets updated to ACTIVE?

File: protocol/testing/e2e/gov/add_new_market_test.go [hunk 25]
```diff
@@ -0,0 +1,256 @@
...
+				&perptypes.MsgCreatePerpetual{
+					Authority: authtypes.NewModuleAddress(govtypes.ModuleName).String(),
+					Params: perptest.GeneratePerpetual(
+						perptest.WithId(5),
+					).Params,
+				},
+				&clobtypes.MsgCreateClobPair{
+					Authority: authtypes.NewModuleAddress(govtypes.ModuleName).String(),
+					ClobPair: *clobtest.GenerateClobPair(
+						clobtest.WithId(5),
```
Comment (ttl33) [inline protocol/testing/e2e/gov/add_new_market_test.go:115]: Should we add a test case where the ID here doesn't exist but is NOT valid? Like 1002

File: protocol/testing/e2e/gov/add_new_market_test.go (hunk 17 above)
Comment (teddyding) [reply protocol/testing/e2e/gov/add_new_market_test.go:55]: Yes

File: protocol/testutil/app/gov.go (hunk 7 above)
Comment (teddyding) [reply protocol/testutil/app/gov.go:58]: Yes

File: protocol/testing/e2e/gov/add_new_market_test.go (hunk 18 above)
Comment (teddyding) [reply protocol/testing/e2e/gov/add_new_market_test.go:95]: We can, but then we are making assumption on the index of the `MsgDelayMessage` in the array. So IMO it's cleaner and more explicit to define this

File: protocol/testing/e2e/gov/add_new_market_test.go (hunk 25 above)
Comment (teddyding) [reply protocol/testing/e2e/gov/add_new_market_test.go:115]: `1002` is a valid ID.

File: protocol/testutil/app/app.go (hunk 21 above)
Comment (teddyding) [reply protocol/testutil/app/app.go:713]: Commented. Needed for submitting transaction that require gas fees (e.g. gov proposals)

File: protocol/testing/e2e/gov/add_new_market_test.go (hunk 24 above)
Comment (teddyding) [reply protocol/testing/e2e/gov/add_new_market_test.go:120]: Added
